lib/livebook/runtime/evaluator.ex

```diff
@@ -688,20 +710,93 @@ defmodule Livebook.Runtime.Evaluator do
     {result, code_markers}
   end
 
-  defp eval(:erlang, code, binding, env) do
+  # Erlang code is either statements as currently supported, or modules.
+  # In case we want to support modules - it makes sense to allow users to use
+  # includes, defines and thus we use the epp-module first - try to find out
+  #
+  # if in the tokens from erl_scan we find at least 1 module-token we assume
+  # that the user is defining a module, if not the previous code is called.
+
+  defp eval(:erlang, code, binding, env, tmp_dir) do
+    case :erl_scan.string(String.to_charlist(code), {1, 1}, [:text]) do
+      {:ok, tokens, _} ->
+        case :lists.keyfind(:module, 3, tokens) do
+          {:atom, _, _module} -> eval_erlang_module(code, binding, env, tmp_dir)
+          false -> eval_erlang_statements(code, tokens, binding, env)
+        end
+
+      {:error, {location, module, description}, _end_loc} ->
+        process_erlang_error(env, code, location, module, description)
+    end
+  end
+
+  # Explain to user: without tmp_dir to write files, they cannot compile erlang-modules
+  defp eval_erlang_module(_code, _binding, _env, nil) do
+    {{:error, :error, "erlang module compile needs tmp_dir", []}, []}
+  end
+
+  # Create module - tokens from string
+  # Based on: https://stackoverflow.com/questions/2160660/how-to-compile-erlang-code-loaded-into-a-string
+  # Step 1: Scan the code using the epp:parse_file erlang primitive
+  # - epp will do macro expansion, includes, defines etc.
+  # Step 2: Compile and load
+  # Step 3: If compile success - register module
+
+  defp eval_erlang_module(code, binding, env, tmp_dir) do
+    # It is an ugly hack for now - need to solve https://github.com/erlang/otp/issues/7239
+    filename = String.to_charlist(tmp_filename(tmp_dir))
+    :file.write_file(filename, code)
+
+    eval_result =
+      case :epp.parse_file(filename, [], []) do
+        {:ok, forms} ->
+          case :lists.keyfind(:module, 3, forms) do
+            {:attribute, _lineno, :module, module} ->
+              case :compile.forms(forms) do
+                {:ok, _, binary_module} ->
+                  {:module, module} = :code.load_binary(module, ~c"#{module}.beam", binary_module)
+
+                  # Registration of module
+                  env = %{env | module: module, versioned_vars: %{}}
+
+                  Evaluator.Tracer.trace({:on_module, binary_module, %{}}, env)
+                  result = {:ok, module}
+                  {{:ok, result, binding, env}, []}
+
+                :error ->
+                  {{:error, :error, "compile forms error", []}, []}
+
+                {:error, _errors, _warnings} ->
+                  {{:error, :error, "compile forms error (warnings)", []}, []}
```

Comment:
It returns only either of these, depending on whether we pass `:return_errors` or not. If the plan is to do it in a separate PR, let's leave a TODO :)

```suggestion
                # TODO: compile with :return_errors, :return_warnings and convert
                # them to diagnostics
                :error ->
                  {{:error, :error, "compile forms error", []}, []}
```